Classification/LibLinear/src/main/java/org/tribuo/classification/liblinear/LibLinearClassificationModel.java

```diff
@@ -291,44 +289,25 @@ protected LibLinearClassificationModel copy(String newName, ModelProvenance newP
 
     @Override
     public OnnxMl.ModelProto exportONNXModel(String domain, long modelVersion) {
-        ONNXContext context = new ONNXContext();
+        ONNXContext onnx = new ONNXContext();
 
-        // Make inputs and outputs
-        OnnxMl.TypeProto inputType = ONNXUtils.buildTensorTypeNode(new ONNXShape(new long[]{-1,featureIDMap.size()}, new String[]{"batch",null}), OnnxMl.TensorProto.DataType.FLOAT);
-        OnnxMl.ValueInfoProto inputValueProto = OnnxMl.ValueInfoProto.newBuilder().setType(inputType).setName("input").build();
-        context.addInput(inputValueProto);
-        OnnxMl.TypeProto outputType = ONNXUtils.buildTensorTypeNode(new ONNXShape(new long[]{-1,outputIDInfo.size()}, new String[]{"batch",null}), OnnxMl.TensorProto.DataType.FLOAT);
-        OnnxMl.ValueInfoProto outputValueProto = OnnxMl.ValueInfoProto.newBuilder().setType(outputType).setName("output").build();
-        context.addOutput(outputValueProto);
-        context.setName("Classification-LibLinear");
+        onnx.setName("Classification-LibLinear");
+        ONNXPlaceholder input = onnx.floatInput(featureIDMap.size());
+        ONNXPlaceholder output = onnx.floatOutput(outputIDInfo.size());
 
         // Build graph
-        writeONNXGraph(context, inputValueProto.getName(), outputValueProto.getName());
-
-        // Build model
-        OnnxMl.ModelProto.Builder builder = OnnxMl.ModelProto.newBuilder();
-        builder.setGraph(context.buildGraph());
-        builder.setDomain(domain);
-        builder.setProducerName("Tribuo");
-        builder.setProducerVersion(Tribuo.VERSION);
-        builder.setModelVersion(modelVersion);
-        builder.setDocString(toString());
-        builder.addOpsetImport(ONNXOperators.getOpsetProto());
-        builder.setIrVersion(6);
-
-        // Extract provenance and store in metadata
-        OnnxMl.StringStringEntryProto.Builder metaBuilder = OnnxMl.StringStringEntryProto.newBuilder();
-        metaBuilder.setKey(ONNXExportable.PROVENANCE_METADATA_FIELD);
-        metaBuilder.setValue(serializeProvenance(getProvenance()));
-        builder.addMetadataProps(metaBuilder.build());
-
-        return builder.build();
+        writeONNXGraph(input).assignTo(output);
```

Comment:
That was the original impetus, but on balance I prefer this approach. I think that passing an object to a method that is essentially what that method is supposed to return is an antipattern - we have a left-hand side for a reason and its not like we're passing around pointers. It requires a function to know more about what will be done with the results of its evaluation than is necessary. 

Now, I've basically replicated this function in a constrained way with `assignTo`, but I think the fact that it doesn't do anything else (and that we need to solve this problem somehow) justifies it. An alternative (perhaps more elegant) approach would be to have an `assign` method only on `ONNXPlaceholder` so that we could write `output.assign(writeONNXGraph(input))`, but I just thought of that just now. That would make it 'feel' more like normal assignment, since the left-hand side is actually on the left. Let me know if you want me to make that change.

Reply:
I agree it is generally an anti-pattern, but in this case we need to know the output variable to properly construct the graph without adding additional identity nodes.

That said, I think that this is nicer than it was before so we can leave that till later.